ADR/0022-secret-mgmt-for-user-workloads.md

```diff
@@ -26,65 +29,142 @@ Accepted
+
+
+#### Example: If the working Secret should be created along with RemoteSecret.
 ```yaml
-apiVersion: v1
-kind: Secret
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
 metadata:
-  name: tmp-jdbc-connection-parameter-upload
-  labels:
-    spi.appstudio.redhat.com/upload-secret: secret
-    spi.appstudio.redhat.com/target-environment: prod
-    spi.appstudio.redhat.com/target-secret-name: jdbc-connection-parameter
-type: Opaque
-data:
- ...
+    namespace: jdoe-tenant
+spec:
+    target:
+        - namespace: spi-test-target
+          secretName: test-remote-secret-secret-5bmq9
+status:
+    TBD
 ```
+> :warning: **Note for this and following:** `RemoteSecret` is not Ready, `SecretData` referenced to it has to be uploaded. See `Uploading secret data to RemoteSecret` example.
+
+#### Example: If RemoteSecret has to be created and uploaded without setting any target
+
 ```yaml
 apiVersion: appstudio.redhat.com/v1beta1
 kind: RemoteSecret
 metadata:
-  name: jdbc-connection-parameter-prod
+    namespace: jdoe-tenant
 spec:
-  secret:
-    environment: prod
-    name: jdbc-connection-parameter
- ...
+    target:
+status:
+  TBD
 ```
 
-#### Example: If the destination is `Namespace` and secret has to be linked to SA
+#### Example:If multiple working Secrets should be created along with RemoteSecret.
 ```yaml
-apiVersion: v1
-kind: Secret
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
 metadata:
-  name: jdbc-connection-parameter
-  annotations:
-      spi.appstudio.redhat.com/secret-linkedto-serviceAccount-name: sa-star
-  labels:
-    spi.appstudio.redhat.com/upload-secret: secret
-    spi.appstudio.redhat.com/target-namespace: ns-9459039345
-type: Opaque
-data:
- ...
+    name: test-remote-secret-secret
+spec:
+    target:
+        - namespace: jdoe-test-ns
+          secretName: test-remote-secret-secret-5bmq9
+        - namespace: jdoe-prod-ns
+          secretName: test-remote-secret-secret-5bmq9
+status:
+    TBD
+```
+
+
+#### Example: If RemoteSecret has to be created with multiple target namespaces including BYC use-case
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+    name: test-remote-secret-secret
+spec:
+    target:
+        - namespace: jdoe-test-ns
+          secretName: test-remote-secret-secret-5bmq9
+        - namespace: jdoe-prod-ns
+          secretName: test-remote-secret-secret-5bmq9
+          apiURL: https://somedomain.copm:443
+          clusterCredentialsSecret: team-a--prod-dtc--secret
+status:
+    TBD
 ```
+
+#### Example: If RemoteSecret has to be created with target namespace and Environment
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+    name: test-remote-secret-secret
+    labels:
+        appstudio.redhat.com/remotesecret-target-environment: prod
+        appstudio.redhat.com/remotesecret-target-component: m-service
+        appstudio.redhat.com/remotesecret-target-application: coffee-shop
+spec:
```

Comment:
Do we need to specify a k8s Secret Type somewhere?

Reply:
@skabashnyuk 
> Now the question of changing the type of an already uploaded secret, it's an interesting point to consider. If someone decides to change the type of an already uploaded secret, the controller should check if the new type is compatible with the existing one.

I am not sure how valid it is in general. 
The question of "compatibility" is not coming from the secret format, secret is just a transport to bring info to the business logic. For example: user application which uses uname/password env variables does not care about the secret bringing it. So, I really doubt compatibility validation can be performed on RS framework level.
To me the choice looks like:
- RS always trying to sync the secrets with RS applying the changes with no conditions
- RS.spec.secret is immutable i.e. RS may change it's state to "unsynced" (for ex) but not try to sync existed secrets (I like this option more for the time - it looks more fair and straightforward)

Reply:
>     RS.spec.secret is immutable i.e. RS may change it's state to "unsynced" (for ex) but not try to sync existed secrets (I like this option more for the time - it looks more fair and straightforward)

This is indistinguishable from the case where the user updates the secret that is the target of the remote secret. In that case, we IMHO should update the secret to conform to the spec.

All in all, I think we should be "dumb" in this regard - we just keep the target secret in sync with the `spec.secret` of the RS. If this means changing the type of the secret, so be it. It may render the data of the secret unusable in the target (because the entries `data` may be incompatible with the new secret type) but that is "not our problem" so to speak. We just reconcile the state.

I think not having the secret spec in the RS but only mirror what we get in the upload secret is a very non-standard thing to do, because there would be no trace of what the target secret of the RS should look like anywhere in the CRD.

On contrary, I would go as far as saying that we should enhance the CRD with a "set of allowed/enforced data fields" so that the spec of the RS documents what kind of secret it is going to create in as much detail as possible. We could also add some smarts to this with marking the RS as invalid if the secret type and allowed fields were incompatible.